src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs

```diff
@@ -69,102 +67,79 @@ private unsafe static long GetAsciiStringAsLong(string str)
-            char* output = stackalloc char[length];
-
-            return GetAsciiStringImplementation(output, input, inputOffset, length);
-        }
+            var asciiString = new string('\0', length);
 
-        private static unsafe string GetAsciiStringImplementation(char* output, byte[] input, int inputOffset, int length)
-        {
-            for (var i = 0; i < length; i++)
+            fixed (char* outputStart = asciiString)
             {
-                output[i] = (char)input[inputOffset + i];
-            }
-
-            return new string(output, 0, length);
-        }
-
-        private static unsafe string GetAsciiStringStack(MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
-        {
-            // avoid declaring other local vars, or doing work with stackalloc
-            // to prevent the .locals init cil flag , see: https://github.com/dotnet/coreclr/issues/1279
-            char* output = stackalloc char[length];
-
-            return GetAsciiStringImplementation(output, start, end, inputOffset, length);
-        }
-
-        private unsafe static string GetAsciiStringHeap(MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
-        {
-            var buffer = new char[length];
-
-            fixed (char* output = buffer)
-            {
-                return GetAsciiStringImplementation(output, start, end, inputOffset, length);
-            }
-        }
-
-        private static unsafe string GetAsciiStringImplementation(char* output, MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
-        {
-            var outputOffset = 0;
-            var block = start;
-            var remaining = length;
-
-            var endBlock = end.Block;
-            var endIndex = end.Index;
+                var output = outputStart;
+                var block = start;
+                var remaining = length;
 
-            while (true)
-            {
-                int following = (block != endBlock ? block.End : endIndex) - inputOffset;
+                var endBlock = end.Block;
+                var endIndex = end.Index;
 
-                if (following > 0)
+                while (true)
                 {
-                    var input = block.Array;
-                    for (var i = 0; i < following; i++)
+                    int following = (block != endBlock ? block.End : endIndex) - inputOffset;
+
+                    if (following > 0)
                     {
-                        output[i + outputOffset] = (char)input[i + inputOffset];
+                        fixed (byte* blockStart = block.Array)
+                        {
+                            var input = blockStart + inputOffset;
+                            var i = 0;
+                            var followingMinusSpan = following - 3;
+                            for (; i < followingMinusSpan; i += 4)
```

Comment:
I would like to say something insightful; but I used a value from the bcl that was probably written in 2000; but cpus have changed since then. While trying to find it I've noted that newer unrolls are much larger e.g. https://github.com/dotnet/coreclr/blob/28468b8f45930632f4a866f50b7b136d85bee5e2/src/mscorlib/src/System/String.cs#L376

Should I do some further benchmarking; get some newer values?